packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,235 @@
+    res.statusCode = 400
+    res.end('"url" parameter is not allowed')
+    return { finished: true }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
+
+  const width = parseInt(w, 10)
+
+  if (!width || isNaN(width)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) must be a number greater than 0')
+    return { finished: true }
+  }
+
+  if (!sizes.includes(width)) {
+    res.statusCode = 400
+    res.end(`"w" parameter (width) of ${width} is not allowed`)
+    return { finished: true }
+  }
+
+  const quality = parseInt(q)
+
+  if (isNaN(quality) || quality < 1 || quality > 100) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) must be a number between 1 and 100')
+    return { finished: true }
+  }
+
+  const { href } = absoluteUrl
+  const hash = getHash([CACHE_VERSION, href, width, quality, mimeType])
+  const imagesDir = join(distDir, 'cache', 'images')
+  const hashDir = join(imagesDir, hash)
+  const now = Date.now()
+
+  if (await fileExists(hashDir, 'directory')) {
+    const files = await promises.readdir(hashDir)
+    for (let file of files) {
+      const [filename, extension] = file.split('.')
+      const expireAt = Number(filename)
+      const contentType = getContentType(extension)
+      if (now < expireAt) {
+        if (contentType) {
+          res.setHeader('Content-Type', contentType)
+        }
+        createReadStream(join(hashDir, file)).pipe(res)
+        return { finished: true }
+      } else {
+        await promises.unlink(join(hashDir, file))
+      }
+    }
+  }
+
+  const upstreamRes = await fetch(href)
+
+  if (!upstreamRes.ok) {
+    server.logError(
+      new Error(`Unexpected status ${upstreamRes.status} from upstream ${href}`)
+    )
+    return { finished: true }
+  }
+
+  const upstreamBuffer = Buffer.from(await upstreamRes.arrayBuffer())
+  const upstreamType = upstreamRes.headers.get('Content-Type')
+  const maxAge = getMaxAge(upstreamRes.headers.get('Cache-Control'))
+  const expireAt = maxAge * 1000 + now
+  let contentType: string
+
+  if (mimeType) {
+    contentType = mimeType
+  } else if (upstreamType?.startsWith('image/') && getExtension(upstreamType)) {
+    contentType = upstreamType
+  } else {
+    contentType = JPEG
+  }
+
+  if (!sharp) {
+    // Lazy load per https://github.com/vercel/next.js/discussions/17141
+    // eslint-disable-next-line import/no-extraneous-dependencies
+    sharp = require('sharp')
```

Comment:
Should we wrap this in a try/catch to show a friendly error if `sharp` isn't installed?

Reply:
Good point, I added a friendly error message.

I also updated `sharp` since a new version came out today and removed the hack in the yarn v2 ci test.